sw/device/lib/dif/dif_uart.h

```diff
@@ -0,0 +1,284 @@
+} dif_uart_enable_t;
+
+/**
+ * Generic odd/even enumeration
+ *
+ * Enumeration used to convey parity information
+ */
+typedef enum dif_uart_parity {
+  kDifUartOdd = 0, /**< odd */
+  kDifUartEven,    /**< even */
+} dif_uart_parity_t;
+
+/**
+ * UART configuration data
+ *
+ * The fundamental data used to configure an UART instance
+ */
+typedef struct dif_uart_config {
+  uint32_t baudrate;               /**< Requested baudrate */
+  uint32_t clk_freq_hz;            /**< Input clock frequency */
+  dif_uart_enable_t parity_enable; /**< Enable parity check */
+  dif_uart_parity_t parity;        /**< Set parity (even by default) */
+} dif_uart_config_t;
+
+/**
+ * UART instance state
+ *
+ * UART persistent data that is required by all UART API. |base_address| must
+ * be initialised by the caller, before passing into the UART DIF init routine.
+ */
+typedef struct dif_uart {
+  reg32_t base_addr; /**< UART base address */
+} dif_uart_t;
+
+/**
+ * Initialise an instance of UART
+ *
+ * Initialise UART instance using the configuration data in @p arg2.
+ * Information that must be retained, and is required to program UART must be
+ * stored in @p arg3.
+ * @param base_addr Base address of an instance of UART IP block
+ * @param config UART configuration data
+ * @param uart UART state data
+ * @return true if the function was successful, false otherwise
+ */
+bool dif_uart_init(reg32_t base_addr, const dif_uart_config_t *config,
+                   dif_uart_t *uart);
+
+/**
+ * Configure an instance of UART
+ *
+ * Configure UART using the configuration data in @p arg2. This operation
+ * performs fundamental configuration.
+ *
+ * @param uart UART state data
+ * @param config UART configuration data
+ * @return true if the function was successful, false otherwise
+ */
+bool dif_uart_configure(const dif_uart_t *uart,
+                        const dif_uart_config_t *config);
+
+/**
+ * Set the RX FIFO watermark
+ *
+ * Set the RX FIFO watermark, is only useful when the corresponding interrupt
+ * is enabled. When the queued RX FIFO number of bytes rises to or above this
+ * level, the RX watermark interrupt is raised.
+ *
+ * @param uart UART state data
+ * @param watermark RX FIFO watermark
+ * @return true if the function was successful, false otherwise
+ */
+bool dif_uart_watermark_rx_set(const dif_uart_t *uart,
+                               dif_uart_watermark_t watermark);
+/**
+ * Set the TX FIFO watermark
+ *
+ * Set the TX FIFO watermark, is only useful when the corresponding interrupt
+ * is enabled. When the queued TX FIFO number of bytes falls to or below this
+ * level, the TX watermark interrupt is raised.
+ *
+ * @param uart UART state data
+ * @param watermark TX FIFO watermark
+ * @return true if the function was successful, false otherwise
+ */
+bool dif_uart_watermark_tx_set(const dif_uart_t *uart,
+                               dif_uart_watermark_t watermark);
+
+/**
+ * Transmit a single UART byte (polled)
```

Comment:
The semantics of this function are a bit strange to me.  I would expect a polled TX operation to return once the byte had been written, but this function waits until the previous TX is done and then enqueues a new TX and returns without completing it.

The semantics of "return when I have sent the byte" allow you to write logic like "send this message and then return".